[PyTorch Numeric Suite] Add support for dynamic LSTM

hx89 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40065 [PyTorch Numeric Suite] Add support for dynamic LSTM**

Add support for dynamic LSTM of all three Numeric Suite APIs: compare_weights(), compare_model_stub() and compare_model_outputs().

Differential Revision: [D22058275](https://our.internmc.facebook.com/intern/diff/D22058275/)

Commit:
[PyTorch Numeric Suite] Add support for dynamic LSTM

Add support for dynamic LSTM of all three Numeric Suite APIs: compare_weights(), compare_model_stub() and compare_model_outputs().

Differential Revision: [D22058275](https://our.internmc.facebook.com/intern/diff/D22058275/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add support for dynamic LSTM"


Add support for dynamic LSTM of all three Numeric Suite APIs: compare_weights(), compare_model_stub() and compare_model_outputs().

Differential Revision: [D22058275](https://our.internmc.facebook.com/intern/diff/D22058275/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add support for dynamic LSTM"


Add support for dynamic LSTM of all three Numeric Suite APIs: compare_weights(), compare_model_stub() and compare_model_outputs().

Differential Revision: [D22058275](https://our.internmc.facebook.com/intern/diff/D22058275/)

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -337,7 +359,10 @@ def compare_model_stub(
     """
     prepare_model_with_stubs(float_model, q_model, module_swap_list, Logger)
-    q_model(data)
+    if type(data) is tuple:
```
We should always pass all the inputs to the model. Why is there a need to have a check here?

hx89 replied:
This is because LSTM module takes two inputs not tuple, so if data is tuple we need to unpack it into two inputs by doing *data.

raghuramank100 replied:
I am still wondering if something simpler is possible: Here is an example:

https://www.internalfb.com/intern/anp/view/?id=285847

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -426,7 +451,11 @@ def compare_model_outputs(
     prepare_model_outputs(float_model, q_model, Logger, white_list)
-    float_model(data)
-    q_model(data)
+    if type(data) is tuple:
```
Similar comment

hx89 replied:
Replied in above comment.

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -211,12 +227,18 @@ def __init__(self, q_module, float_module, Logger):
-        shadow_output = self.shadow_module(x)
-        self.logger(output, shadow_output)
+    def forward(self, *x):
+        if len(x) > 1:
```
Shadow module should pass all arguments to orig and shadow_module. The logger should also take in the entire outputs. Internally the logger could choose to log output[0], at the forward here, lets keep it general.

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -93,6 +93,22 @@ def compare_weights(float_dict, quantized_dict):
+            print('lstm float hh key is: ', float_weight_hh_key)
+            if float_weight_ih_key in float_dict and float_weight_hh_key in float_dict:
+                weight_dict[key] = {}
+                weight_dict[key]["float"] = float_dict[float_weight_ih_key]
```
This feels hacky, I am going to write a PR to make it easy to get weights for LSTMs

hx89 replied:
Can we land this PR first to have a working prototype? Once you made the changes I can update it to use the new way to get weights.

raghuramank100 reviewed (changes requested):
Great work!, requesting some changes,

Commit:
Update on "[PyTorch Numeric Suite] Add support for dynamic LSTM"


Add support for dynamic LSTM of all three Numeric Suite APIs: compare_weights(), compare_model_stub() and compare_model_outputs().

Differential Revision: [D22058275](https://our.internmc.facebook.com/intern/diff/D22058275/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add support for dynamic LSTM"


Add support for dynamic LSTM of all three Numeric Suite APIs: compare_weights(), compare_model_stub() and compare_model_outputs().

Differential Revision: [D22058275](https://our.internmc.facebook.com/intern/diff/D22058275/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add support for dynamic LSTM"


Add support for dynamic LSTM of all three Numeric Suite APIs: compare_weights(), compare_model_stub() and compare_model_outputs().

Differential Revision: [D22058275](https://our.internmc.facebook.com/intern/diff/D22058275/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add support for dynamic LSTM"


Add support for dynamic LSTM of all three Numeric Suite APIs: compare_weights(), compare_model_stub() and compare_model_outputs().

Differential Revision: [D22058275](https://our.internmc.facebook.com/intern/diff/D22058275/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add support for dynamic LSTM"


Add support for dynamic LSTM of all three Numeric Suite APIs: compare_weights(), compare_model_stub() and compare_model_outputs().

Differential Revision: [D22058275](https://our.internmc.facebook.com/intern/diff/D22058275/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add support for dynamic LSTM"


Add support for dynamic LSTM of all three Numeric Suite APIs: compare_weights(), compare_model_stub() and compare_model_outputs().

Differential Revision: [D22058275](https://our.internmc.facebook.com/intern/diff/D22058275/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add support for dynamic LSTM"


Add support for dynamic LSTM of all three Numeric Suite APIs: compare_weights(), compare_model_stub() and compare_model_outputs().

Differential Revision: [D22058275](https://our.internmc.facebook.com/intern/diff/D22058275/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add support for dynamic LSTM"


Add support for dynamic LSTM of all three Numeric Suite APIs: compare_weights(), compare_model_stub() and compare_model_outputs().

Differential Revision: [D22058275](https://our.internmc.facebook.com/intern/diff/D22058275/)

[ghstack-poisoned]
